### PR TITLE
Add support for open and read timeouts

### PR DESCRIPTION
**Pull request by jlaine**
This is an alternative take on #316 using Python's time.time().

### CODE REVIEW

File: av/container/core.pyx [hunk 1]
```diff
@@ -150,6 +164,12 @@ cdef class ContainerProxy(object):
             # Finish releasing the whole structure.
             lib.avformat_free_context(self.ptr)
 
+    def __set_callback_timeout__(self, timeout):
+        self.callback_info.timeout = timeout
+
+    def __reset_start_time__(self):
+        self.callback_info.start_time = time.time() * 1000
```
[the remark this answers is not in the document]
Comment (jlaine) [reply av/container/core.pyx]: `start_time` is an int64_t, no cast is needed

**Comment by jlaine** [comment]
I have made a significant change : after the open / read operation completes, we clear the timeout. There may be other places where the `AVIOInterruptCb` is used and we don't want to cause breakage.

File: av/container/core.pyx [hunk 2]
```diff
@@ -203,6 +229,11 @@ cdef class Container(object):
         if sentinel is not _cinit_sentinel:
             raise RuntimeError('cannot construct base Container')
 
+        timeouts = options.pop('timeouts', {}) if options else {}
+        timeouts = timeouts if timeouts is not None else {}
+        self.open_timeout = int(timeouts.get('open_timeout', 30000))
```
Comment (jlaine) [inline av/container/core.pyx]: I'm not sure I'm comfortable with using timeouts by default. @mikeboers any thoughts?

**Comment by jlaine** [comment]
@mikeboers if we set the defaults to "0" (no timeout) would you be comfortable with merging this?

**Comment by mikeboers** [comment]
I've not read it, and don't think I can today, but I'm happy with it with no changes to default behavior if you're ready to merge it.

_takes a second to read it anyways_

I'm a little iffy on overloading `options` for it, but I'm game. If we hate it later we can support it for a couple major versions. NBD.

Ship it!

**Comment by jlaine** [comment]
Actually I think I agree with you @mikeboers, "options" is already confusing enough as it is. Maybe explicit `open_timeout` and `read_timeout` defaulting to `None` would be better?

**Comment by mikeboers** [comment]
That would be better. Maybe it is confusing, but [Requests' `request.request` `timeout`](http://docs.python-requests.org/en/master/api/#requests.request) is interesting.

(It is a float in which case it is used for both, or a 2-tuple for connect and read timeouts.)

**Comment by jlaine** [comment]
The `requests` example is indeed interesting. Regardless of whether we use a tuple or two values I think using seconds (and not milliseconds) like requests is more pythonic.

**Comment by mikeboers** [comment]
It should absolutely be seconds.

**Comment by jlaine** [comment]
The timeouts are now in seconds. I'd still like to add some unit tests, but I haven't come up with a good testcase yet. I'll try spinning up an HTTP server in a child process and make it reply slooowly.

**Comment by mikeboers** [comment]
FYI I'm totally sold on the requests model. Most of the time you just want a timeout, and don't care to set them separately.

**Comment by jlaine** [comment]
OK I have changed the API to take a single `timeout` argument as requested.

I have also managed to add a test for the "open timeout", which led me to realize my code was badly broken: storing a time as a C `float` is useless, we need a `double`.

**Comment by mikeboers** [comment]
I've restarted the two failed Travis builds. AppVeyor doesn't like it at all.

CI may not like it, but I do.

**Comment by jlaine** [comment]
The Travis failure looks like flakiness, which I can probably work on.

The AppVeyor failure on the other hand looks like something else altogether. The test which is failing doesn't use timeouts at all, opening the file should just work.

What's this error?

```
copying build\lib.win32-2.7\av\buffer.pyd -> av
error: could not delete 'av\buffer.pyd': Access is denied
```

**Comment by jlaine** [comment]
CI is in a very weird state right now:

- on Travis, random unrelated tests are hanging

- on AppVeyor, switching from "multiprocessing" to "threading" has eliminated the "Access is denied" failure but the new tests are hanging

**Comment by mikeboers** [comment]
This is passing having been rebased onto develop and using the new testing infrastructure (which is still missing PyPy).

**Comment by mikeboers** [comment]
I take it back. It failed on one of the windows machines. LOL

**Pull request merged**